### PR TITLE
FEAT: Parallel upload of files to CAS

### PR DESCRIPTION
**Pull request by heynemann**
With this feature we can do parallel uploads to Hyper-CAS.

Adding @cezarguimaraes and @diegoximenes here as reviewers just for a sanity check before merging this to master.

### CODE REVIEW

File: utils/lock.go [hunk 1]
```diff
@@ -9,23 +9,22 @@ import (
 
 // Possibility for the future https://medium.com/@gdm85/distributed-locking-for-pennies-distrilock-967347e7f2dd
 
-var NoOp = func() error { return nil }
+var noOp = func() error { return nil }
 
+// Lock resource if viper returns it should be locked.
 func Lock(resource string) (func() error, error) {
-	viper.SetDefault("file.enableLocks", true)
```
Comment (heynemann) [inline utils/lock.go:15]: These are not thread-safe, that's why I removed them.

**Review by diegoximenes — APPROVED** [review]
LGTM!

Added a few comments that don't block a merge.

However the next comment should be considered for a future PR.
Today all files' contents are first read into memory and then distributed to workers.
Memory could be better used if only the file's paths were distributed to workers, and then each worker would be responsible to read a file and process it.

Also, it seems that the deploying cycle of `hyper-cas serve` and `hyper-case sync` are independent, therefore would be nice to see them in separate repositories.

To deploy this in our infrastructure is just a matter of creating a tag in GitHub, and deploying this new tag in our control plane cluster with ioctl.

File: synchronizer/core.go [hunk 2]
```diff
@@ -53,53 +76,107 @@ func listFiles(path string) ([]string, []string, error) {
 	return files, contents, err
 }
 
-func NewSync(root, apiURL string) *Sync {
-	return &Sync{rootDir: root, apiURL: apiURL}
+func initHTTPClient(retryCount, maxConcurrentRequests, httpTimeoutMs int) *httpclient.Client {
```
[the remark this answers is not in the document]
Comment (heynemann) [reply synchronizer/core.go]: Now I've added two flags, one for each purpose. It makes sense to have them both, and now it should be clearer. Thanks for the catch!

File: synchronizer/core.go [hunk 3]
```diff
@@ -53,53 +76,107 @@ func listFiles(path string) ([]string, []string, error) {
...
+		s.wg.Add(1)
+		go s.worker(s.jobChan, s.respChan, &s.wg)
+	}
+}
+
+func (s *Sync) worker(jobChan chan *fileUpdateJob, respChan chan *fileUpdateResponse, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	for job := range jobChan {
+		hash, isUpToDate, err := s.uploadFile(job.path, job.content)
```
[the remark this answers is not in the document]
Comment (heynemann) [reply synchronizer/core.go:133]: Path is useful for debugging in that function, and it is a private function. But I agree with that the `isUpToDate` change is sound!

File: storage/fs.go [hunk 4]
```diff
@@ -129,6 +136,11 @@ func (st *FSStorage) StoreDistro(root string, hashes []string) error {
 	finalPath := path.Join(st.sitesPath, root)
 	err = os.Rename(dir, finalPath)
 	if err != nil {
+		sErr, ok := err.(*os.LinkError)
+		if ok && sErr.Op == "rename" {
+			utils.LogWarn("Distribution path already exist. Ignoring rename...", zap.Error(err), zap.String("path", finalPath))
+			return nil
+		}
```
Comment (heynemann) [inline storage/fs.go:143]: @cezarguimaraes the change you suggested!

**Review by diegoximenes — APPROVED** [review]
LGTM

Just added some comments that can be ignored.

**Pull request merged**